README.md
# smenu
Smenu is a set of hierachical scripts to manage Oracle database


scripts/history.txt
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
#                 visit  "www.smenu.org" for latest Smenu release                        +
#+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
10-05-2023  :   Altered sq -m for additional detail on waits.
10-05-2023  :   Added sq -eh for detailed information on query performance corelating db events.
14-03-2023  :   Altered asm -f to reflect non mirrored free space
13-12-2022  :   Added sts -adv' to list advisors status
07-12-2022  :   Added 'rdl -la -i <inst>' to distinguish on which node the redo activity took place
23-11-2022  :   Added 'aw -writes' to display IO & direct writes stats in conjuction with '-sl' '-each'
21-11-2022  :   Added 'soc -s' 'soc -curr' 'soc -cpt' 'soc -prg' to investigate open cursors
                when you hit 'too many open cursors'
19-10-2022  :   Added tbl -unload -cvs, to unload data with  "," as separator
14-09-2022  :   Added 'cb' and 'cf' to change color background and font
10-08-2022  :   Added sl -la and sl -ltop to display session info from v$active_session_history
25-07-2022  :   Added column everage rows to sq -ph
15-07-2022  :   Added sts -awlb to list meta for Kruger MANAGE_PARTS
01-07-2022  :   Added profile to "gc -bk" to list backup info from cloud control
20-06-2022  :   Added profile to "usr -cr <oneuser>" when profile is default
02-02-2022  :   added sta -def to view dbms_gather_Stats default setting
28-12-2021  :   added 'lstd -gap f <file_id>' to list size & gap in a datafile. lstd -s accept
                now -f <file_id> to calculate resize only for a single datafile.
30-09-2021  :   Added 'lstd -fsa' to list filenumber for temporary tablespace, case of ORA-01114 
                for temporary datafile it is the filenumber and not the file id that is reported
10-06-2021  :   Enriched aud -ls -at -u with activity stats
12-02-2021  :   Added dep -lsv List all grants tables needed for views of a user
21-01-2021  :   Added idx -lcl to list orphaned entries in indexes to be cleanup
04-06-2019  :   Added idx -use to list SQL_ID using this index
03-05-2019  :   Added src -ext <view> to xpand sub view text to top level (only 12c+)
30-04-2019  :   Added tbl -tis (concat of tbl -t, tbl -t -s, idx -t). nice for analysis
25-04-2019  :   Added sx -purge to remove a given SID plan from SGA
02-04-2019  :   Added 'sq -mcount' to list bind mistatch query count and reason
20-02-2019  :   Modified sq -top to list sql_id instead of hash_value
20-02-2019  :   Added lsbk -jbb to list last 10 db backups and last 10 arch back. This is to
                avoid missing the db backup entries due to spam from numerous arch backup
18-12-2018  :   Added aud -ls -c <code> to see failed loging attempt (code=1017)
04-12-2018  :   Added sts -nls to see NLS setting
25-10-2017  :   Added  tbl -rdef -pm -col_name -seg to generated a redef script to partition
                a table on -pm months on column -col_name with -seg sub partition per month
20-10-2017  :   Added  aw -each to single list snap_id  between -b and -e
08-09-2017  :   Added  lc -purge SQL_ID to remove SQL_ID from library cache
08-02-2017  :   Added  rol -tx <obj> to list excution grants on procedures
29-12-2016  :   Added 'aw -sts' to list statistics variance over 2 weeks
11-10-2016  :   Added 'tbl tsiz -tl <list of tables>  to process a list of table wastage, 
                blank separated
26-09-2016  :   Added 'tbl -hist' to derive a table history growth from AWR.
25-08-2016  :   Added 'aw -rdl' to list LGWR and REDO statistics
25-07-2016  :   Added 'soc -cpt' to list the count of opened cursors per session
09-06-2016  :   Added 'aw -hist -t <tbl>' to list stts from WRI\$_OPTSTAT_TAB_HISTORY
21-03-2016  :   added 'sta -lst' to list statle statistics for non system tables
16-03-2016  :   added 'aw -pga' to list stats of pga from AWR
23-12-2015  :   added 'sq -fm' to find query with same signature and different litterals
22-12-2015  :   added 'sx -lh' to list the hints of a stored profile
21-12-2015  :   added 'sts -ltb' to list admin_tab_partitions configuration
21-12-2015  :   added 'tbl -rdef' to create script to online alter a table using dbms_redefinition
08-12-2015  :   Added 'sq -le' to list sql_exec_id. Added 'sq -ex' to display sql_monitor
                report for a given exec_id
02-10-2015  :   Added change background color in 'cnt' if the SID is listed in the new file $SBINS/.prod
03-09-2015  :   Added shortcut 'gc' to deal with some unsupported Grid control features.
                added 'gc -td' to list target DB. Added 'gc -rmw' to remove warning messages.
                added 'gc -rmc' to remove critical messages
03-08-2015  :   Added 'aw -slp' to list session history under crash cicumstances (more ligt query)
                added 'drm -lw' to list maintenance windows of the resource manager
26-02-2015  :   Added 'frg -rcb' to display content of recycle bin
10-12-2014  :   Added 'sts -al -ora', 'sts -cpt','sts -bid <n>' show alert log excerpt
                and counts.
27-11-2014  :   Added 'frg -aux' to list aux occupancy
26-11-2014  :   Added 'frg -tmp' to list session temp occupancy
14-11-2014  :   Added 'usr -lck' to view lcoked users, 'usr -xp' to view expired, 'usr -rst'
                to generate a new passwrd a the text of a mail to used by the DBA (RRE)
15-10-2014  :   Added 'sta -set_freq' to set a frequency histogram columns
06-10-2014  :   Added 'spx -dop', 'spx -f', 'spx -o' to show info over parallel executions
25-08-2014  :   Added mgm shortcut to deal with Grid control metrics.
22-07-2014  :   Added 'dg -r' to show apply log statistics 
20-05-2014  :   Added sga -top to list active occupancy of sql work area
13-02-2014  :   Added shortcut 'met' to deal with system metric
05-02-2014  :   Added shortcut 'buf -lo -t <table>' to list all buffer associated to all 
                objects related to the table in to the cache

scripts/tkp.pl
#!/usr/bin/perl 
#===============================================================================
#
#         FILE:  10046_events.pl
#
#        USAGE:  ./10046_events.pl  
#
#  DESCRIPTION: Provides statistical report of 10046 wait events from
#               raw trace file.
#
#               For pre Oracle 10.2 trace files the breakdown and histogram
#               sections of the report are based on unique P1 and event
#               combinations. For tracefiles produced from 10.2 onwards
#               these sections are reported for each unique combination 
#               of obj# and event.
#
#      OPTIONS:  -h -t trace_file.trc
# REQUIREMENTS:  ---
#         BUGS:  ---
#        NOTES:  ---
#       AUTHOR:  Clive Bostock (Oracle ACS)
#      COMPANY:  
#      VERSION:  1.2
#      CREATED:  17/04/2011 11:14:22
#     REVISION:  19/05/2011 Kyle Hailey (Oracle ACE) - modified parsing event names 
#===============================================================================
use strict;
use warnings;
use File::Basename;
use Getopt::Std;

my @trace_file;
my @events;
my %event_records;
my %tot_sort;
my %tot_objn_sort;
my $ver = '1.2';
my $trace_file;
our $opt_t;
our $opt_h;
our $prog     = basename($0);
my $event_line;
my $event_rec;
my $aggr_rec;
my $er;
my %event_stat;
my %event_aggr;
my $event;
my $elapsed;
my $tot_ela = 0;
my $ela_ms;
my $objid;
my $objid_event;
my $pre_10_2    = 0;
my $obj_header  = 'Object Id';
my $e1    = 0;
my $e2    = 0;
my $e4    = 0;
my $e8    = 0;
my $e16   = 0;
my $e32   = 0;
my $e64   = 0;
my $e128  = 0;
my $e256  = 0;
my $e512  = 0;
my $e1024 = 0;

our $SD="\/";
our $SP=":";
our $os = $ENV{'OS'}; 
if ( ! $os )
  { $os = $ENV{'OSTYPE'}; }
if ( ! $os )
  { 
     $os = `uname`;
     chomp $os;  
  }
if ( ! $os )
  { 
     $os = 'Unknown';
  }

$SD="\\" if($os  =~ /Win/i);
$SP=";" if($os   =~ /Win/i);

getopts('ht:') or 
    die "\nInvalid options specified, use $prog -h.\n$prog: Deploying chute and bailing out!!!\n";

if (defined($opt_t))
{
    $trace_file = $opt_t;
}
if (defined($opt_h))
{
    print "\nUsage:  $prog -t trace_file.trc\n\n";
    exit;
}

if (! defined($opt_t))
{
    print "\nUsage:  $prog -t trace_file.trc\n\n";
    exit;
}

open(TRACEFILE, $trace_file) || die "Failed to open trace file ($trace_file)\n";

printf "$prog: 10046 traces events analyser ver $ver\n\n";
@trace_file = <TRACEFILE>;
close(TRACEFILE);

print "*** Trace file header details ***\n\n";
foreach $event_line (@trace_file)
{
    chomp $event_line;
    if (length($event_line))
         { printf("$event_line\n"); }
    else { last; }
}
print "Trace input file : $trace_file\n";
                                       #**************************************/
                                       #* Load only the WAIT event lines     */
                                       #* into @events                       */
                                       #**************************************/
@events = grep(/WAIT/, @trace_file);
chomp @events;

foreach $event_line (@events)
{

    $event = $elapsed = $objid = $event_line;
    # $event =~ s/^.*nam='([\\:\w\s\d-]+)'.*/$1/;
    $event =~ s/^.*nam='//;
    $event =~ s/' ela=.*//;
    $elapsed  =~ s/^.*ela=[\s]*([\d]+) .*/$1/;
    if ( $pre_10_2 == 0 )
    {
        if ( $event_line !~ m/obj#/ )
        { 
            print "\n\n***********************************************\n";
            print "*** Trace file format looks pre Oracle 10.2 ***\n";
            print "***********************************************\n\n";
            $pre_10_2 = 1; 
        }
    }

    if ($pre_10_2 )
    {
        $objid  =~ s/^.*p1=[-\s]*([\d]+) .*/$1/;
        $obj_header = 'P1';
    }
    else
    {
        $objid  =~ s/^.*obj#=[-\s]*([\d]+) .*/$1/;
    }

    $objid_event = sprintf("%-19s: %-45s", $objid, $event);
#   printf ("%s %s %s\n", $event, $elapsed, $objid);

                                       #**************************************/
                                       #*  We maintain this (simple) hash    */
                                       #* as a method of sorting by total    */
                                       #* elapsed time per event             */
                                       #**************************************/
    if (exists $tot_sort{$event})
    {    
      $tot_sort{$event} = $tot_sort{$event} + $elapsed;
    }
    else
    {
      $tot_sort{$event} = $elapsed;
    }

                                       #**************************************/
                                       #*  Get histogram counts              */
                                       #**************************************/
    $ela_ms = $elapsed / 1000;
    $tot_ela = $tot_ela + $elapsed;
    $e1 = $e2 = $e4 = $e8 = $e16 = $e32 = $e64 = $e128  = $e256 = $e512 = $e1024 = 0;
    if ( $ela_ms < 1 )
    { $e1 = 1 }
    elsif ( $ela_ms < 2  &&  $ela_ms >= 1 )
    { $e2 = 1 }
    elsif ( $ela_ms < 4  &&  $ela_ms >= 2 )
    { $e4 = 1 }
    elsif ( $ela_ms < 8  &&  $ela_ms >= 4 )
    { $e8 = 1 }
    elsif ( $ela_ms < 16  &&  $ela_ms >= 8 )
    { $e16 = 1 }
    elsif ( $ela_ms < 32  &&  $ela_ms >= 16 )
    { $e32 = 1 }
    elsif ( $ela_ms < 64  &&  $ela_ms >= 32 )
    { $e64 = 1 }
    elsif ( $ela_ms < 128  &&  $ela_ms >= 64 )
    { $e64 = 1 }
    elsif ( $ela_ms < 256  &&  $ela_ms >= 128 )
    { $e128 = 1 }
    elsif ( $ela_ms < 512  &&  $ela_ms >= 256 )
    { $e256 = 1 }
    else 
    { $e1024 = 1 };


                                       #**************************************/
                                       #*  Build an anonymous record         */
                                       #* structure containing our event     */
                                       #* data.                              */
                                       #**************************************/
    $event_rec = {OBJ_EVENT => $objid_event
               , TOT_ELA => $elapsed
               , COUNT => 1
               , E1    => $e1
               , E2    => $e2
               , E4    => $e4
               , E8    => $e8
               , E16   => $e16
               , E32   => $e32
               , E64   => $e64
               , E128  => $e128
               , E256  => $e256
               , E512  => $e512
               , E1024 => $e1024
                 };
    if ($er = $event_stat{$objid_event})
        { 
            $er->{COUNT}   = $er->{COUNT} + 1;
            $er->{TOT_ELA} = $er->{TOT_ELA} + $event_rec->{TOT_ELA};
            $er->{TOT_ELA} = $er->{TOT_ELA} + 1;
            $er->{E1}      = $er->{E1}      + $event_rec->{E1};
            $er->{E2}      = $er->{E2}      + $event_rec->{E2};
            $er->{E4}      = $er->{E4}      + $event_rec->{E4};
            $er->{E8}      = $er->{E8}      + $event_rec->{E8};
            $er->{E16}     = $er->{E16}     + $event_rec->{E16};
            $er->{E32}     = $er->{E32}     + $event_rec->{E32};
            $er->{E64}     = $er->{E64}     + $event_rec->{E64};
            $er->{E128}    = $er->{E128}    + $event_rec->{E128};
            $er->{E256}    = $er->{E256}    + $event_rec->{E256};
            $er->{E512}    = $er->{E512}    + $event_rec->{E512};
            $er->{E1024}   = $er->{E1024}   + $event_rec->{E1024};
            $event_stat {$er -> {OBJ_EVENT}} = $er;
            $er->{E1} = $er->{E1} + $event_rec->{E1};
            $tot_objn_sort{$objid_event} = $tot_objn_sort{$objid_event} + $elapsed;
        }
    else
        { 
            $event_stat {$event_rec -> {OBJ_EVENT}} = $event_rec;
            $tot_objn_sort{$objid_event} = $elapsed;
        }

    $aggr_rec = {EVENT => $event, TOT_ELA => $elapsed,  COUNT => 1};
    if ($er = $event_aggr{$event})
        { 
            $er->{COUNT}   = $er->{COUNT} + 1;
            $er->{TOT_ELA} = $er->{TOT_ELA} + $aggr_rec->{TOT_ELA};
            $er->{TOT_ELA} = $er->{TOT_ELA} + 1;
            $event_aggr {$er -> {EVENT}} = $er;
        }
    else
        { 
            $event_aggr {$aggr_rec -> {EVENT}} = $aggr_rec;
        }
}

print "\nEVENT AGGREGATES\n";
print "================\n\n";
printf ("Wait Event %s Count        Elapsed (ms)     Avg Ela (ms)  %%Total \n",' ' x 34);
printf ("~~~~~~~~~~~%s ~~~~~~~~~~~~ ~~~~~~~~~~~~     ~~~~~~~~~~~~  ~~~~~~\n",'~' x 34);
foreach $objid_event (sort {$tot_sort{$b} <=>  $tot_sort{$a} } keys %event_aggr)
{
    my $ev = $event_aggr{$objid_event};
    printf("%45s %12d %12d     %12d  %6.2f\n",$ev->{EVENT}, $ev->{COUNT}, $ev->{TOT_ELA}/1000, ($ev->{TOT_ELA}/$ev->{COUNT})/1000, ($ev->{TOT_ELA}*100/$tot_ela) ); 
}
printf("           %s              ~~~~~~~~~~~~                         \n",' ' x 34);
printf ("         %s Total Elapsed: %12d                                   \n",' ' x 34,$tot_ela / 1000);
print "\nEVENT AGGREGATE BREAKDOWN\n";
print "=========================\n\n";
printf ("%-9s          : Wait Event %s Count        Tot Ela (ms) %%Total Avg Ela (ms)\n",$obj_header, ' ' x 34);
printf ("~~~~~~~~~~~~~~~~~~ : ~~~~~~~~~~~%s ~~~~~~~~~~~~ ~~~~~~~~~~~~ ~~~~~~ ~~~~~~~~~~~~\n",'~' x 34);
foreach $objid_event (sort {$tot_objn_sort{$b} <=>  $tot_objn_sort{$a} } keys %event_stat)
{
    my $ev = $event_stat{$objid_event};
    printf("%57s %12d %12d  %5.2f %12d\n",$ev->{OBJ_EVENT}, $ev->{COUNT}, $ev->{TOT_ELA}/1000
                                       , $ev->{TOT_ELA}*100/$tot_ela, ($ev->{TOT_ELA}/$ev->{COUNT})/1000); 
}
print "\nEVENT HISTOGRAM BREAKDOWN\n";
print "===========================\n\n";
print "This section splits the event counts into elapsed time\n";
print "buckets so that we can see if there are any suspiciousn\n";
print "or anomalous response time / frequency patterns.\n\n";
printf("%-9s          : Wait Event %s <1ms    <2ms    <4ms    <8ms    <16ms   <32ms   <64ms   <128ms  <256ms  <512ms  >=1024ms\n",$obj_header, ' ' x 34);
printf("~~~~~~~~~~~~~~~~~~ : ~~~~~~~~~~~%s ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~ ~~~~~~~~\n",'~' x 34);
foreach $objid_event (sort {$tot_objn_sort{$b} <=>  $tot_objn_sort{$a} } keys %event_stat)
{
    my $ev = $event_stat{$objid_event};
    printf("%60s ", $ev->{OBJ_EVENT});
    printf("%7s %7s %7s ", $ev->{E1},    $ev->{E2},    $ev->{E4}); 
    printf("%7s %7s %7s ", $ev->{E8},    $ev->{E16},   $ev->{E32});
    printf("%7s %7s %7s ", $ev->{E64},   $ev->{E128},  $ev->{E256});
    printf("%7s %8s\n", $ev->{E512},  $ev->{E1024} ); 
}
printf("\n\n*** End of report ***\n");

